gamemode/modules/party/sv_party.lua

```diff
@@ -0,0 +1,359 @@
+    [2] = "invite only",
+}
+
+util.AddNetworkString( "RVR_Party_updateClient" )
+util.AddNetworkString( "RVR_Party_createParty" )
+util.AddNetworkString( "RVR_Party_updateFriends" )
+
+local function updateClientPartyData( id )
+    local partyData = party.getParty( id )
+
+    net.Start( "RVR_Party_updateClient" )
+        net.WriteUInt( id, 32 )
+        net.WriteBool( tobool( partyData ) )
+        if partyData then
+            net.WriteTable( partyData )
+        end
+    net.Broadcast()
+end
+
+function party.createParty( partyName, owner, tag, color, joinMode )
+    for id, partyData in pairs( party.parties ) do
+        if partyData.name == partyName then
+            return nil, "Party with name " .. partyName .. " already exists"
+        end
+        if partyData.tag == tag then
+            return nil, "Party with tag " .. tag .. " already exists"
+        end
+    end
+
+    if owner:GetPartyID() then
+        return nil, "Player " .. owner:Nick() .. " already in a party"
+    end
+
+    if #tag ~= 4 then
+        return nil, "Tag must be 4 characters"
+    end
+
+    local minLen = GAMEMODE.Config.Party.MIN_PARTY_NAME_LENGTH
+    local maxLen = GAMEMODE.Config.Party.MAX_PARTY_NAME_LENGTH
+    if #partyName < minLen or #partyName > maxLen or tonumber( partyName ) then
+        return nil, "Invalid party name, must be between " .. minLen .. " and " ..
+            maxLen .. " characters and contain at least one non-numeric character"
+    end
+
+    if color.a ~= 255 then
+        return nil, "Party color alpha must be 255 (ff)"
+    end
+
+    local id = party.idCounter
+    party.idCounter = party.idCounter + 1
+    local partyData = {
+        name = partyName,
+        owner = owner,
+        members = { owner },
+        id = id,
+        tag = tag,
+        color = color,
+        joinMode = joinMode,
+        invites = {},
+    }
+
+    party.parties[id] = partyData
+
+    owner:SetPartyID( id )
+
+    hook.Run( "RVR_Party_PartyCreated", partyData )
+
+    updateClientPartyData( id )
+
+    return id, partyData
+end
+
+function party.removeParty( id )
+    local partyData = party.getParty( id )
+
+    if not partyData then
+        return false, "Party with id " .. id .. " does not exist"
+    end
+
+    for _, ply in pairs( partyData.members ) do
+        ply:SetPartyID( nil )
+    end
+
+    party.parties[id] = nil
+
+    updateClientPartyData( id )
+
+    return true
+end
+
+function party.broadcastMessage( id, msg )
+    local partyData = party.getParty( id )
+
+    if not partyData then
+        return false, "Party with id " .. id .. " does not exist"
+    end
+
+    for _, member in pairs( partyData.members ) do
+        member:ChatPrint( msg )
+    end
+
+    return true
+end
+
+function party.setJoinMode( id, mode )
+    local partyData = party.getParty( id )
+
+    if not partyData then
+        return false, "Party with id " .. id .. " does not exist"
+    end
+
+    if mode < 0 or mode > 2 then
```

Comment:
same comment as earlier, use joinmod constants or the length of the joinmode table